gcloud/bigquery/table.py

```diff
@@ -693,6 +699,225 @@ def insert_data(self,
+            else:
+                raise ValueError('total bytes could not be determined. Please '
+                                 'pass an explicit size.')
+        headers = {
+            'Accept': 'application/json',
+            'Accept-Encoding': 'gzip, deflate',
+            'User-Agent': connection.USER_AGENT,
+            'content-type': 'application/json',
+        }
+
+        metadata = {
+            'configuration': {
+                'load': {
+                    'sourceFormat': source_format,
+                    'schema': {
+                        'fields': _build_schema_resource(self._schema),
+                    },
+                    'destinationTable': {
+                        'projectId': self._dataset.project,
+                        'datasetId': self._dataset.name,
+                        'tableId': self.name,
+                    }
+                }
+            }
+        }
+
+        self._configure_job_metadata(metadata, allow_jagged_rows,
+                                     allow_quoted_newlines, create_disposition,
+                                     encoding, field_delimiter,
+                                     ignore_unknown_values, max_bad_records,
+                                     quote_character, skip_leading_rows,
+                                     write_disposition)
+
+        upload = Upload(file_obj, content_type, total_bytes,
+                        auto_transfer=False)
+
+        url_builder = _UrlBuilder()
+        upload_config = _UploadConfig()
+
+        # Temporary URL, until we know simple vs. resumable.
+        base_url = connection.API_BASE_URL + '/upload'
+        path = '/projects/%s/jobs' % self._dataset.project
+        upload_url = connection.build_api_url(api_base_url=base_url, path=path)
+
+        # Use apitools 'Upload' facility.
+        request = Request(upload_url, 'POST', headers,
+                          body=json.dumps(metadata))
+
+        upload.configure_request(upload_config, request, url_builder)
+        query_params = url_builder.query_params
+        base_url = connection.API_BASE_URL + '/upload'
+        request.url = connection.build_api_url(api_base_url=base_url,
+                                               path=path,
+                                               query_params=query_params)
+        upload.initialize_upload(request, connection.http)
+
+        if upload.strategy == RESUMABLE_UPLOAD:
+            http_response = upload.stream_file(use_chunks=True)
+        else:
+            http_response = make_api_request(connection.http, request,
+                                             retries=num_retries)
+        response_content = http_response.content
+        if not isinstance(response_content,
+                          six.string_types):  # pragma: NO COVER  Python3
+            response_content = response_content.decode('utf-8')
+        return client.job_from_resource(json.loads(response_content))
```

Comment:
This code is very much like `Blob.upload_from_file` (not surprising, as I started from there), with the complication that the "job configuration" stuff:  there is nowhere else to set it up, because it  has to be present in the initial request.  If we made all those parameters kwargs, we might reduce the mechanical complexity, but only at the cost of head-scratching when trying to figure out how to use them (what should the keys be, what are the expected types / constraints, etc.)